Examples/Scripts/Python/pythia8.py

```diff
@@ -1,51 +1,173 @@
 #!/usr/bin/env python3
-import os
+from typing import Optional, Union
+from pathlib import Path
+from collections.abc import Iterable
 
 import acts
 import acts.examples
 
-from common import addPythia8
-
 u = acts.UnitConstants
 
 
-def runPythia8(
-    outputDir,
-    outputRoot: bool = True,
-    outputCsv: bool = True,
-    s: acts.examples.Sequencer = None,
-):
+def addPythia8(
+    s: acts.examples.Sequencer,
+    rnd: Optional[acts.examples.RandomNumbers] = None,
+    nhard: int = 1,
+    npileup: int = 200,
+    beam: Optional[
+        Union[acts.PdgParticle, Iterable]
+    ] = None,  # default: acts.PdgParticle.eProton
+    cmsEnergy: Optional[float] = None,  # default: 14 * acts.UnitConstants.TeV
+    hardProcess: Optional[Iterable] = None,  # default: ["HardQCD:all = on"]
+    pileupProcess: Iterable = ["SoftQCD:all = on"],
+    vtxGen: Optional[acts.examples.EventGenerator.VertexGenerator] = None,
+    outputDirCsv: Optional[Union[Path, str]] = None,
+    outputDirRoot: Optional[Union[Path, str]] = None,
+    printParticles: bool = False,
+    returnEvGen: bool = False,
+) -> acts.examples.Sequencer:
+    """This function steers the particle generation using Pythia8
+
+    NB. this is a reimplementation of common.addPythia8, which is maintained for now for compatibility.
+
+    Parameters
+    ----------
+    s: Sequencer
+        the sequencer module to which we add the particle gun steps (returned from addParticleGun)
+    rnd : RandomNumbers, None
+        random number generator
+    nhard, npileup : int, 1, 200
+        Number of hard-scatter and pileup vertices
+    beam : PdgParticle|[PdgParticle,PdgParticle], eProton
+        beam particle(s)
+    cmsEnergy : float, 14 TeV
+        CMS energy
+    hardProcess, pileupProcess : [str], ["HardQCD:all = on"], ["SoftQCD:all = on"]
+        hard and pileup processes
+    vtxGen : VertexGenerator, None
+        vertex generator module
+    outputDirCsv : Path|str, path, None
+        the output folder for the Csv output, None triggers no output
+    outputDirRoot : Path|str, path, None
+        the output folder for the Root output, None triggers no output
+    printParticles : bool, False
+        print generated particles
+    """
```

Comment:
Missing `returnEvGen` description in docstring

Reply:
I'll add:
```
    returnEvGen: bool, False
        returns EventGenerator instead of Sequencer.
        This option  is included for compatibility and will be removed when common.addPythia8 is removed.
```

I mention this now, because this reminds me of a minor, but general issue with splitting our Python examples up into modules. There is no way for the caller to inspect or modify the algorithms/readers/writers, since they are just added to the `Sequencer` and not returned to the caller.

We could do that, but many modules set up several algorithms so we'd have to think a bit how best to do it. But the caller already has the `Sequencer`, which has this information - just no way to inspect it as far as I could easily tell. Could we look at a way of querying the algorithms, readers, writers etc? Eg. allowing something like:

```
outputParticles = s.lastEntry(acts.examples.EventGenerator).config.outputParticles
```

`acts.examples.Sequencer.lastEntry(type)` would return the last entry that matches the type. If that's difficult, we could search by name. Actually the C++ would probably require separate separate methods for `lastAlgorithm`, `lastReader`, `lastWriter` etc, so maybe that's better. We could do something more general and complicated, but that would probably suit this use case well.

Sorry for the digression. What do you and @paulgessinger think?

Reply:
PS. related to the above: normally (unless `returnEvgen=True`) we return the `Sequencer`, although this is also passed in by the user. This is perhaps useful to make clear that the the `Sequencer` has been modified, but isn't actually useful.

Would it help to have the examples just call without resetting `s`?

Reply:
BTW, all my discussion is about ideas for the future. I hope it won't delay this PR.

Reply:
I'm not sure. I would try to keep this meta-level somewhat simple, and try to encourage people to copy the actual algorithm constructions into their own scripts if they want to more thoroughly customize. That was kind of the original idea of these basic scripts.

I feel like with this PR we have a good middle ground: use the `addXXX` function if you just want to compose what's already there, and fall back to a "full" script if you want to control more closely what's being set up.

Reply:
> I'll add:
> 
> ```
>     returnEvGen: bool, False
>         returns EventGenerator instead of Sequencer.
>         This option  is included for compatibility and will be removed when common.addPythia8 is removed.
> ```
From the docstring point of view, this is enough for me.

> I mention this now, because this reminds me of a minor, but general issue with splitting our Python examples up into modules. There is no way for the caller to inspect or modify the algorithms/readers/writers, since they are just added to the `Sequencer` and not returned to the caller.
> 
> We could do that, but many modules set up several algorithms so we'd have to think a bit how best to do it. But the caller already has the `Sequencer`, which has this information - just no way to inspect it as far as I could easily tell. 

I also tried to split the Python examples into modules before seeing this PR and my first thought was also to extract info from the algorithm added into the Sequencer but didn't find a way to do so.

Allowing something like:
> 
> ```
> outputParticles = s.lastEntry(acts.examples.EventGenerator).config.outputParticles
> ```
would make easier the chaining of new algorithms. 

> BTW, all my discussion is about ideas for the future. I hope it won't delay this PR.

I think that such change should'nt be done in this PR as it is more related on how to chain new algorithm. As long as everything works with the current algorithm, I would delay that into another PR.

Reply:
I'll resolve this thread when I commit this (and other docstring fixes)

Reply:
> @paulgessinger had suggested maybe later moving the `add*` functions to separate files in the Python module tree. That might be a good time to consider this sort of thing.

Yes, I think that examples should be kept simple and not callable, something ```full_chain_itk.py```-like.